### PR TITLE
Don't lose your heads over infinite loops

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Bugfixes "NPCs faint when infinite looping, instead of dying."```

#### Purpose of change
Despite it being a constant and slightly humorous part of developing NPCs and adding new bugs to them, it really isnt *needed* that NPCs heads explode when they are infinite looping.
This is a leftover from the very early days of the game when NPCs werent as important.

#### Describe the solution
Ive added a reboot() function to NPCs that resets a lot of their AI state, resets their activities, goal points, path points, destinations , everything, then wacks them over the head with a metaphysical hammer, so they faint into a coma.

This is preferable to killing them.

Now they are asleep, they dont process anymore, and hopefully the reboot has fixed their brain, so you can talk to them and tell them to wake up, hopefully they are fixed, if not, theyll immediately faint again, and you can keep them in a coma while you wait for a bugfix, instead of losing your beloved NPC forever in a cranial conflagration.

This should cover most causes of infinite loops, with the list of things that are reset/cleared

#### Describe alternatives you've considered
N/A

#### Testing
Deliberately induced an infinite loop with an activity that didnt subtract moves at all.
NPC fainted, got debugmsg as usual, told them to wake up, they were fine again, as they no longer had an activity. 

#### Additional context
1_sNnuIM29rLqCRUgKwyvYuA

### CODE REVIEW

**Comment by Kodiologist** [comment]
Don't forget to update the comment about "cranial detonation". Similarly, I don't think the new comment "instead of destroying NPC" makes sense because it refers to an aspect of the code that will no longer exist once this PR is merged.

File: src/game.cpp [hunk 1]
```diff
@@ -4264,8 +4267,8 @@ void game::monmove()
         // If we spun too long trying to decide what to do (without spending moves),
         // Invoke cranial detonation to prevent an infinite loop.
         if( turns == 10 ) {
-            add_msg( _( "%s's brain explodes!" ), guy.name );
-            guy.die( nullptr );
+            add_msg( _( "%s's faints!" ), guy.name );
```
Comment (esotericist) [inline src/game.cpp]:
```suggestion
            add_msg( _( "%s faints!" ), guy.name );
```

File: src/game.cpp [hunk 2]
```diff
@@ -4264,8 +4267,8 @@ void game::monmove()
         // If we spun too long trying to decide what to do (without spending moves),
         // Invoke cranial detonation to prevent an infinite loop.
```
Comment (esotericist) [inline src/game.cpp]:
```suggestion
        // Invoke cognitive suspension to prevent an infinite loop.
```

**Comment by esotericist** [comment]
> Similarly, I don't think the new comment "instead of destroying NPC" makes sense because it refers to an aspect of the code that will no longer exist once this PR is merged.

I think having a mention that this was the historical practice isn't without merit there, but I'm personally not particularly concerned either way.

**Comment by esotericist** [comment]
As an additional thought: is there merit in giving the player a pop-up before talking to an NPC that was suspended, to let them know/remind them that they're trying to wake up an NPC that was suspended due to an infinite loop?

**Comment by Kodiologist** [comment]
You could certainly say something like "A previous version of the game just destroyed the NPC in this case", but the current wording suggests that destroying the NPC is somehow a possibility. You can direct the reader to check previous commits if they're interested, but you can't rely on them already knowing how the code used to be.

**Comment by davidpwbrown** [comment]
> As an additional thought: is there merit in giving the player a pop-up before talking to an NPC that was suspended, to let them know/remind them that they're trying to wake up an NPC that was suspended due to an infinite loop?

"Hi, I'm in a debug coma, want to try and wake me up to see if the problem was fixed?"

or do we try and keep some form of immersion ? and pretend its like fainting debug goat disease, idk.

**Comment by esotericist** [comment]
> > As an additional thought: is there merit in giving the player a pop-up before talking to an NPC that was suspended, to let them know/remind them that they're trying to wake up an NPC that was suspended due to an infinite loop?
> 
> "Hi, I'm in a debug coma, want to try and wake me up to see if the problem was fixed?"
> 
> or do we try and keep some form of immersion ? and pretend its like fainting debug goat disease, idk.

I don't think immersion is necessary here, given that an error is a prerequisite.

my slightly frazzled brain wants to suggest something like:
```
This character was previously put into a suspension state due to an AI bug.
If you have reason to believe the problem was fixed, you can wake them up.
Do you want to do that now?
```

but I know that's wordy. at the least, I think this isn't a place where we need to try to be cutesy, and should instead be clear about what's going on.

**Comment by adamkad1** [comment]
'Your npc has finished rebooting, do you want to reactivate them?' <that would be amusing

**Pull request merged**